appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.web.loader;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import org.apache.naming.resources.FileDirContext;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class WebappClassLoaderTest {
+
+    private static CyclicBarrier lock;
+    private static ExecutorService executor;
+    private static File junitJarFile;
+
+    @BeforeClass
+    public static void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        lock = new CyclicBarrier(3);
+        executor = Executors.newFixedThreadPool(60);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @AfterClass
+    public static void shutdown() {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final ClassLoader classLoader = this.getClass().getClassLoader();
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(classLoader, null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Exception> result = new CompletableFuture<>();
+
+        // Create the tasks, and have them each run at the same time
+        // using the cyclic barrier
+        Runnable lookupTask = waitAndDo(lock, result, () -> lookup(classLoader, webappClassLoader));
+        Runnable addTask = waitAndDo(lock, result, () -> add(classLoader, webappClassLoader));
+        Runnable closeTask = waitAndDo(lock, result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < 100; i++) {
+                executor.execute(addTask);
```

Comment:
LABEL for cycle with executor: note that the pool uses a queue, so first three tasks will get their time and then all other (do I remember it well?), until the pool will be empty. Following requests will wait in the queue for free (finished) threads.